db/listings.go

```diff
@@ -0,0 +1,734 @@
+	err := json.Unmarshal(messageText, &message)
+	if err != nil {
+		return nil, errors.Wrap(err, wrapMsg)
+	}
+
+	// Update any fields that we need to update.
+	message.Type = notificationType
+	message.Seen = seen
+	message.Deleted = deleted
+
+	return &message, nil
+}
+
+// V1NotificationListingParameters describes the parameters available for listing notifications.
+type V1NotificationListingParameters struct {
+	User             string
+	Offset           uint64
+	Limit            uint64
+	Seen             *bool
+	SortOrder        query.SortOrder
+	SortField        query.V1ListingSortField
+	NotificationType string
+}
+
+// getNotificationListingSortColumn returns the sort column to use for a V1ListingSortField value.
+func getV1NotificationListingSortColumn(sortField query.V1ListingSortField) (string, error) {
+	switch sortField {
+	case query.V1ListingSortFieldTimestamp, query.V1ListingSortFieldDateCreated:
+		return "n.time_created", nil
+	case query.V1ListingSortFieldUUID:
+		return "n.id", nil
+	case query.V1ListingSortFieldSubject:
+		return "n.subject", nil
+	}
+	return "", fmt.Errorf("unrecognized sort field: %s", string(sortField))
+}
+
+// V1ListNotifications lists notifications for a user.
+func V1ListNotifications(tx *sql.Tx, params *V1NotificationListingParameters) (*model.V1NotificationListing, error) {
+	wrapMsg := "unable to obtain the notification listing"
+
+	// Begin building the query.
+	queryBuilder := sq.StatementBuilder.
+		PlaceholderFormat(sq.Dollar).
+		Select().
+		Column("nt.name AS type").
+		Column("n.seen").
+		Column("n.deleted").
+		Column("n.outgoing_json AS message").
+		Column("count(*) OVER () AS total").
+		From("notifications n").
+		Join("users u ON n.user_id = u.id").
+		Join("notification_types nt ON n.notification_type_id = nt.id").
+		Where(sq.Eq{"u.username": params.User}).
+		Where(sq.Eq{"n.deleted": false})
+
+	// Apply the seen parameter if requested.
+	if params.Seen != nil {
+		queryBuilder = queryBuilder.Where(sq.Eq{"n.seen": *params.Seen})
+	}
+
+	// Apply the notification type parameter if requested.
+	if params.NotificationType != "" {
+		queryBuilder = queryBuilder.Where(sq.Eq{"nt.name": params.NotificationType})
+	}
+
+	// Apply the limit if requested.
+	if params.Limit > 0 {
+		queryBuilder = queryBuilder.Limit(params.Limit)
+	}
+
+	// Apply the offset if requested.
+	if params.Offset != 0 {
+		queryBuilder = queryBuilder.Offset(params.Offset)
+	}
+
+	// Apply sorting.
+	sortColumn, err := getV1NotificationListingSortColumn(params.SortField)
+	if err != nil {
+		return nil, errors.Wrap(err, wrapMsg)
+	}
+	queryBuilder = queryBuilder.OrderBy(fmt.Sprintf("%s %s", sortColumn, string(params.SortOrder)))
+
+	// Build the query.
+	query, args, err := queryBuilder.ToSql()
+	if err != nil {
+		return nil, errors.Wrap(err, wrapMsg)
+	}
+
+	// Query the database.
+	rows, err := tx.Query(query, args...)
+	if err != nil {
+		return nil, errors.Wrap(err, wrapMsg)
+	}
+	defer rows.Close()
```

Comment:
I've tended to avoid the `Run` and `RunWith` primarily because I wanted to have a way to easily add a statement to log a query without having to modify the code much. The extra verbosity felt worthwhile to me for that reason. Admittedly, I didn't spend any time looking through squirrel to see if it has a built-in way to log the query.

Reply:
After glancing over the code in squirrel, I think we can create a type that implements the `BaseRunner` interface and wrap the transaction when we want to log a query. That would allow us to pass the transaction in directly in cases where the query should not be logged and pass the transaction wrapped with a new `LoggingBaseRunner` type that logs the query and arguments before passing them to the transaction:

``` go
rows, err := queryBuilder.RunWith(newLoggingBaseRunner(tx)).Query()
if err != nil {
    return nil, errors.Wrap(err, wrapMsg)
}
defer rows.Close()
```

We could get fancier by making it a service-wide configuration setting, but that would require more refactoring, and logging every query generated by the service would probably get overwhelming anyway. I don't think I'm going to make this change now, but it does seem like it might be useful.

`squirrel.go` has a function called `DebugSqlizer` that also sounds like it might be useful. I haven't had time to look into it in detail yet, though.

Reply:
Yeah, what you're saying looks like what I see as well, looking at squirrel. It appears that `DebugSqlizer` is more like a way to produce a single string for debugging rather than producing `query` and `args` and having to combine them yourself. The `newLoggingBaseRunner` approach seems interesting, especially since it could probably be made to pass in a logging level. All that being said, your reasoning for not using `Run`/`RunWith` seems pretty sensible to me, so I'm happy either way.

Reply:
Oh, I randomly noticed another option here -- for debugging, it can still be pretty easy if you do, say:

```go
s, args, _ := query.ToSql()
log.Info(s, args)

rows, err := query.RunWith(...).Query()
if err != nil { ...etc }
```

basically, the first `err` can be ignored since it'd come up again in the second one, so those two lines are sufficient to log the query/args, without needing to do two full error blocks.

May or may not be to your taste, just figured I should mention it since I thought of it!